### PR TITLE
Add support for bf16/fp32 mix precision in cuda_batch_norm

### PR DESCRIPTION
**Pull request by robert1003**
Fixes #ISSUE_NUMBER

* CUDNN's batch norm kernel supports bf16 input + fp32 weight input ([docs](https://docs.nvidia.com/deeplearning/cudnn/api/cudnn-ops-library.html#cudnnbatchnormalizationbackwardEx)) but pytorch does not. This PR allows it.
* Fix a bug in constant type. Constants (alpha and beta) should have the same type as the weights not the input.

### CODE REVIEW

**Review by eqy — COMMENTED** [review]
Should tests be added/updated to test this on GPUs >= sm80?

**Comment by robert1003** [comment]
Hi @eqy yes that make sense, could you let me know where is the test file I should modify and how should I run the test? Sorry I am not really familiar with the overall codebase...

**Comment by robert1003** [comment]
Got it thanks will do it this weekend!

**Comment by robert1003** [comment]
@eqy I added a test in `test/test_nn.py` to test batchnorm in cudnn + bf16 setting. I think the mixed_precision test is failing because of the bug I found, which I have fixed in subsequent commits. Could you take a look? Thanks!

**Review by eqy — APPROVED** [review]

**Comment by robert1003** [comment]
@eqy It seems that some of the errors in CIs are not caused by my changes:
```
/var/lib/jenkins/workspace/aten/src/ATen/cuda/CUDAContextLight.h:6:10: fatal error: cuda_runtime_api.h: No such file or directory
    6 | #include <cuda_runtime_api.h>
```

Is this expected? Thanks!

**Comment by eqy** [comment]
Might need a rebase after resolving merge conflicts; you can do this with `@pytorchmergebot rebase`

**Comment by robert1003** [comment]
I think the compile error is caused by trying to import `#include <ATen/cuda/CUDAContext.h>` for cpu machines. Is there a flag for coda such that I can use for `#ifdef <FLAG>`? 

Thank you!

**Comment by robert1003** [comment]
@eqy I'm not sure if `#ifdef USE_CUDA` is the correct switch to use...

File: aten/src/ATen/native/Normalization.cpp [hunk 1]
```diff
@@ -58,6 +58,10 @@
 #include <ATen/ops/sqrt.h>
 #endif
 
+#ifdef USE_CUDA
+#include <ATen/cuda/CUDAContext.h>
```
Comment (eqy) [inline aten/src/ATen/native/Normalization.cpp:62]: My understanding is that `Normalization.cpp` shouldn't depend on `CUDAContext.h`; the typical way that this kind of support check is done is via `CUDAHooks.cpp` which can query things like the device properties if CUDA is available---`Normalization.cpp` can call into `CUDAHooks` without needing to depend on anything CUDA-specific. You can see examples of other support-checking functions https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/cuda/detail/CUDAHooks.cpp

**Comment by robert1003** [comment]
@eqy Is this PR ready to merge?

File: test/test_nn.py [hunk 2]
```diff
@@ -4897,6 +4897,26 @@ def test_batchnorm_cudnn_half(self):
             self.assertEqual(cudnn_output, thnn_output)
             self.assertEqual(cudnn_input_grad, thnn_input_grad, atol=1e-3, rtol=0)
 
+    @unittest.skipIf(not TEST_CUDA, "CUDA unavailable")
+    def test_batchnorm_cudnn_bf16(self):
+        # THNN
+        input = torch.randint(1, 10, (2, 3, 2, 2), dtype=torch.bfloat16, device="cuda", requires_grad=True)
```
Comment (eqy) [inline test/test_nn.py:4903]: How is this choosing which backend to dispatch to? Would we need a context manager or something else to prevent it from going to cuDNN to generate the reference?

**Comment by eqy** [comment]
@robert1003 I think it looks mostly okay, see the one comment on how the reference is generated for the test

Otherwise please rebase so the failing tests can be rerun

**Comment by robert1003** [comment]
I just update my branch with the newest commit on PyTorch master!

@pytorchmergebot rebase

**Comment by robert1003** [comment]
@eqy I don't think I have the permission to use the rebase bot...
